optimizer/lib/transformers/BlurImagePlaceholders.js

```diff
@@ -0,0 +1,158 @@
+/**
+ * Copyright 2018 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+const sizeOf = require('image-size');
+const jimp = require('jimp');
+
+/**
+ * Adds placeholders for all AMP images and posters that are blurry versions of
+ * the corresponding original source. The blur will be displayed as the
+ * <amp-img> is rendering, and will fade out once the element is loaded.
+ */
+class BlurImagePlaceholders {
+  /**
+   * Parses the document to add blurred placedholders in all appropriate
+   * locations.
+   * @param tree A parse5 treeAdapter.
+   * @return {Array} An array of promises that all represents the resolution of
+   * a blurred placeholder being added in an appropriate place.
+   */
+  transform(tree) {
+    const html = tree.root.firstChildByTag('html');
+    const body = html.firstChildByTag('body');
+    const promises = [];
+    for (let node = body; node !== null; node = node.nextNode()) {
+      const {tagName} = node;
+      let src;
+      if (tagName === 'amp-img') {
+        src = node.attribs.src;
+      }
+      if (tagName === 'amp-video' && node.attribs.poster) {
+        src = node.attribs.poster;
+      }
+      if (src && !this.hasPlaceholder_(node)) {
```

Comment:
Do we need to check for the `noloading` attribute or is this handled by the AMP runtime? 

Should we also check for a minimum image size? I think the blur effect looks distracting for small icons and logos.

Reply:
Assuming my understanding of noloading is correct, I think checking for it here would be good regardless due to the smaller size of the AMP page that is produced by the tool.

Also, yes minimum image size will be addressed in a follow up PR!